### PR TITLE
[ISSUE #4058] DLedgerCommitLog support LMQ

### PR DESCRIPTION
**Pull request by chaiyx**
## What is the purpose of the change

#4058

## Brief changelog

XX

## Verifying this change

XXXX

Follow this checklist to help us incorporate your contribution quickly and easily. Notice, `it would be helpful if you could finish the following 5 checklist(the last one is not necessary)before request the community to review your PR`.

- [x] Make sure there is a [Github issue](https://github.com/apache/rocketmq/issues) filed for the change (usually before you start working on it). Trivial changes like typos do not require a Github issue. Your pull request should address just this issue, without pulling in other changes - one PR resolves one issue. 
- [x] Format the pull request title like `[ISSUE #123] Fix UnknownException when host config not exist`. Each commit in the pull request should have a meaningful subject line and body.
- [x] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
- [x] Write necessary unit-test(over 80% coverage) to verify your logic correction, more mock a little better when cross module dependency exist. If the new feature or significant change is committed, please remember to add integration-test in [test module](https://github.com/apache/rocketmq/tree/master/test).
- [x] Run `mvn -B clean apache-rat:check findbugs:findbugs checkstyle:checkstyle` to make sure basic checks pass. Run `mvn clean install -DskipITs` to make sure unit-test pass. Run `mvn clean test-compile failsafe:integration-test`  to make sure integration-test pass.
- [ ] If this contribution is large, please file an [Apache Individual Contributor License Agreement](http://www.apache.org/licenses/#clas).

### CODE REVIEW

File: store/src/main/java/org/apache/rocketmq/store/dledger/DLedgerCommitLog.java [hunk 1]
```diff
@@ -421,6 +421,11 @@ private void setMessageInfo(MessageExtBrokerInner msg, int tranType) {
 
         setMessageInfo(msg, tranType);
 
+        boolean multiDispatchWrapResult = multiDispatch.wrapMultiDispatch(msg);
```
Comment (Git-Yang) [inline store/src/main/java/org/apache/rocketmq/store/dledger/DLedgerCommitLog.java:424]: multiDispatch.wrapMultiDispatch() has a concurrency problem.

**Comment by RongtongJin** [comment]
@tianliuliu Could you help review this pull request?

**Review by RongtongJin — APPROVED** [review]

**Review by Git-Yang — CHANGES_REQUESTED** [review]
image
multiDispatch.wrapMultiDispatch() should be inside lock.

**Comment by tianliuliu** [comment]
> @tianliuliu Could you help review this pull request?

yes，I am glad to be

**Comment by tianliuliu** [comment]
Can you assign it to me? There is a problem with the implementation of this place

File: store/src/main/java/org/apache/rocketmq/store/dledger/DLedgerCommitLog.java (hunk 1 above)
Comment (chaiyx) [reply store/src/main/java/org/apache/rocketmq/store/dledger/DLedgerCommitLog.java]: fixed

**Comment by chaiyx** [comment]
> 

I have put the message serialization method of lmq into the lock, does it fit?

**Comment by chaiyx** [comment]
> multiDispatch.wrapMultiDispatch(msg) can not use here

**Comment by tianliuliu** [comment]
> > 
> 
> @tianliuliu I have put the message serialization method of lmq into the lock, does it fit?

use a new method to implement the part, in wrapMultiDispatch(msg) has encoded the message and There no need

File: store/src/main/java/org/apache/rocketmq/store/dledger/DLedgerCommitLog.java (hunk 1 above)
Comment (tianliuliu) [reply store/src/main/java/org/apache/rocketmq/store/dledger/DLedgerCommitLog.java]: > fixed

the keyBuilder can be ThreadLocal?

**Review by vongosling — APPROVED** [review]

**Comment by duhenglucky** [comment]
ping @chaiyx

**Review by cserwen — APPROVED** [review]
LGTM

**Pull request merged**